components/fxa-client/src/device.rs

```diff
@@ -263,10 +283,20 @@ impl FirefoxAccount {
         self.update_device(builder.build())
     }
 
-    fn update_device(&self, update: DeviceUpdateRequest<'_>) -> Result<UpdateDeviceResponse> {
+    fn update_device(&mut self, update: DeviceUpdateRequest<'_>) -> Result<UpdateDeviceResponse> {
         let refresh_token = self.get_refresh_token()?;
-        self.client
-            .update_device(&self.state.config, refresh_token, update)
+        let res = self
+            .client
+            .update_device(&self.state.config, refresh_token, update);
+        match res {
+            Ok(resp) => Ok(resp),
+            Err(err) => {
+                // We failed to write an update to the server.
+                // Clear local state so that we'll be sure to retry later.
+                self.state.device_capabilities.clear();
+                Err(err)
```

Comment:
@eoger I added state-clearing here on failure. This means we clear it in some cases when we don't technically need to (e.g. if we were only updating the device name) and also means that some other methods now need to take an `&mut self`, but it seemed better on balance to have this done in a single location.

I'd still like to do the broader refactor with a `StateV3` described in the linked issue, but another time.